pkgs/stdenv/darwin/make-bootstrap-tools.nix

```diff
@@ -29,188 +13,243 @@ let cross = if crossSystem != null
+        local lib rpath
         for lib in $libs; do
           ${stdenv.cc.targetPrefix}install_name_tool -change $lib "@rpath/$(basename "$lib")" "$1"
         done
-      }
 
-      # Strip executables even further
-      for i in $out/bin/*; do
-        if [[ ! -L $i ]] && isMachO "$i"; then
-          chmod +w $i
-          ${stdenv.cc.targetPrefix}strip $i || true
+        case "$(dirname "$1")" in
+        */bin)
+          # Strip executables even further
+          ${stdenv.cc.targetPrefix}strip "$i"
+          rpath='@executable_path/../lib'
+          ;;
+        */lib)
+          # the '/.' suffix is required
+          rpath='@loader_path/.'
+          ;;
+        */lib/darwin)
+          rpath='@loader_path/..'
+          ;;
+        *)
+          echo unkown executable $1 >&2
+          exit 1
+          ;;
+        esac
+
+        # if shared object contains references add an rpath to lib/
+        if ${stdenv.cc.targetPrefix}otool -l "$1"| grep -q '@rpath/'; then
+          ${stdenv.cc.targetPrefix}install_name_tool -add_rpath "$rpath" "$1"
         fi
-      done
+      }
 
-      for i in $out/bin/* $out/lib/*.dylib $out/lib/darwin/*.dylib; do
-        if [[ ! -L "$i" ]]; then
-          rpathify $i
-        fi
-      done
+      # check that linked library paths exist in $out/lib
+      # must be run after rpathify is performed
+      checkDeps() {
+        local deps=$(${stdenv.cc.targetPrefix}otool -l "$1"| grep -o '@rpath/[^      ]*' || true)
+        local lib
+        for lib in $deps; do
+          if [[ ! -e $out/''${lib/@rpath/lib} ]]; then
+            echo "error: $1 missing lib for $lib" >&2
+            exit 1
+          fi
+        done
+      }
 
-      for i in $out/bin/*; do
-        if [[ ! -L "$i" ]] && isMachO "$i"; then
-          ${stdenv.cc.targetPrefix}install_name_tool -add_rpath '@executable_path/../lib' $i
+      for i in $out/bin/* unpack/bin/* $out/lib{,/darwin}/*.dylib; do
+        if [[ ! -L $i ]] && isMachO "$i"; then
+          rpathify "$i"
+          checkDeps "$i"
         fi
       done
 
-      ${if stdenv.targetPlatform.isx86_64 then ''
-        rpathify $out/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
-      '' else ''
-        sed -i -e 's|/nix/store/.*/libobjc.A.dylib|@executable_path/../libobjc.A.dylib|g' \
-          $out/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation.tbd
-      ''}
-
+      nuke-refs $out/bin/*
       nuke-refs $out/lib/*
-      nuke-refs $out/lib/system/*
       nuke-refs $out/lib/darwin/*
-      ${lib.optionalString stdenv.targetPlatform.isx86_64 ''
-        nuke-refs $out/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
-      ''}
+      nuke-refs $out/lib/system/*
+      nuke-refs unpack/bin/*
+
+      # remove reference present in aarch64. however `@executable_path/../libobjc.A.dylib`
+      # looks like perhaps should be `@executable_path/../lib/libobjc.A.dylib`
```

Comment:
Cane we check in a built bootstrap-tools whether or not this is the case?

Reply:
this is an artifact and was introduced in a previous change. i added the comment as it looks wrong but i haven't written test code to figure out what the deal is with tbd files. suggestions are welcome.

do i just link to a tbd file like i would a dylib?  i think yes.

Reply:
from some trial and error. when linking to `CoreFoundation.tbd` and modifying the contents of `libraries: ['@executable_path/..//llibobjc.A.dylib']`
- `@executable_path` represents the current directory of the linker
- `@loader_path` represents the current directory of the CoreFoundation.tbd
- if the basename is `libobjc.A.dylib` it doesn't matter what the parent is, link will succeed
- renaming the `libobjc.A.tbd` -> `llibobj.A.tbd` will create linker errors if path is not correct

so using `@loader_path` seems like it is correct except if the file is symlinked then `@loader_path` is relative to the symlink and not the target. which breaks. this is not the case for rpaths in dylibs as `@loader_path` through a symlink always seems to be the location of the target.

anyway, it seems like these files need to be patched with an absolute path.

Reply:
fixed to patch in absolute path. verified path correct in aarch64.

Reply:
was able to remove CF as suggested by @reckenrode which removes the file from the bootstrap tools.